dex/farm-staking-contracts/metabonding-staking/src/lib.rs

```diff
@@ -0,0 +1,101 @@
+#![no_std]
+
+use locked_asset_token::StakingEntry;
+
+elrond_wasm::imports!();
+
+pub mod locked_asset_token;
+
+pub type SnapshotEntry<M> = MultiValue2<ManagedAddress<M>, BigUint<M>>;
+pub const UNBOND_EPOCHS: u64 = 10;
+
+#[elrond_wasm::contract]
+pub trait MetabondingStaking: locked_asset_token::LockedAssetTokenModule {
+    #[init]
+    fn init(
+        &self,
+        locked_asset_token_id: TokenIdentifier,
+        locked_asset_factory_address: ManagedAddress,
+    ) {
+        self.locked_asset_token_id().set(&locked_asset_token_id);
+        self.locked_asset_factory_address()
+            .set(&locked_asset_factory_address);
+    }
+
+    #[payable("*")]
+    #[endpoint(stakeLockedAsset)]
+    fn stake_locked_asset(&self) {
+        let payments = self.call_value().all_esdt_transfers();
+        self.require_all_locked_asset_payments(&payments);
+
+        let caller = self.blockchain().get_caller();
+        let new_locked_asset_token = self.merge_locked_asset_tokens_if_needed(&caller, payments);
+
+        self.total_locked_asset_supply()
+            .update(|total_supply| *total_supply += &new_locked_asset_token.amount);
+
+        self.staking_entry_for_user(&caller).set(&StakingEntry::new(
+            new_locked_asset_token.token_nonce,
+            new_locked_asset_token.amount,
+        ));
+        self.user_list().insert(caller);
+    }
+
+    #[endpoint]
+    fn unstake(&self) {
+        let caller = self.blockchain().get_caller();
+        let entry_mapper = self.staking_entry_for_user(&caller);
+        require!(!entry_mapper.is_empty(), "Must stake first");
+
+        let mut staking_entry: StakingEntry<Self::Api> = entry_mapper.get();
+        require!(!staking_entry.is_unstaked(), "Already unstaked");
+
+        let current_epoch = self.blockchain().get_block_epoch();
+        staking_entry.opt_unbond_epoch = Some(current_epoch + UNBOND_EPOCHS);
+
+        entry_mapper.set(&staking_entry);
+    }
+
+    #[endpoint]
+    fn unbond(&self) {
+        let caller = self.blockchain().get_caller();
+        let entry_mapper = self.staking_entry_for_user(&caller);
+        require!(!entry_mapper.is_empty(), "Must stake first");
+
+        let staking_entry: StakingEntry<Self::Api> = entry_mapper.get();
+        require!(staking_entry.is_unstaked(), "Must unstake first");
+
+        let current_epoch = self.blockchain().get_block_epoch();
+        let unbond_epoch = unsafe { staking_entry.opt_unbond_epoch.unwrap_unchecked() };
```

Comment:
why do you need unsafe block?

Reply:
`unwrap_unchecked` is `unsafe fn`, This if guaranteed to be `Some`, as `staking_entry.is_unstaked()` checks for `staking_entry.opt_unbond_epoch.is_some()`

Reply:
i get your point but... in my opinion... use unwrap, its more simple